contracts/dtoken/src/deposit.rs

```diff
@@ -0,0 +1,167 @@
+use crate::*;
+use std::convert::TryFrom;
+
+const GAS_FOR_DEPOSIT: Gas = Gas(120_000_000_000_000);
+const MARGIN_TRADING_CONTRACT: &str = "mtrading.omomo-finance.testnet";
+
+impl Contract {
+    pub fn post_deposit(&mut self, token_amount: WBalance) -> PromiseOrValue<WBalance> {
+        if !is_promise_success() {
+            return PromiseOrValue::Value(token_amount);
+        }
+        underlying_token::ft_balance_of(
+            env::current_account_id(),
+            self.get_underlying_contract_address(),
+            NO_DEPOSIT,
+            TGAS,
+        )
+        .then(ext_self::deposit_balance_of_callback(
+            token_amount,
+            env::current_account_id(),
+            NO_DEPOSIT,
+            self.terra_gas(50),
+        ))
+        .into()
+    }
+}
+
+#[near_bindgen]
+impl Contract {
+    pub fn deposit(&mut self, token_amount: WBalance) -> PromiseOrValue<WBalance> {
+        require!(
+            env::prepaid_gas() >= GAS_FOR_DEPOSIT,
+            "Prepaid gas is not enough for deposit flow"
+        );
+
+        self.mutex_account_lock(Actions::Deposit, token_amount, self.terra_gas(120))
+    }
+
+    #[private]
+    pub fn deposit_balance_of_callback(
+        &mut self,
+        token_amount: WBalance,
+    ) -> PromiseOrValue<WBalance> {
+        if !is_promise_success() {
+            log!(
+                "{}",
+                Events::DepositFailedToGetUnderlyingBalance(
+                    env::signer_account_id(),
+                    Balance::from(token_amount),
+                    self.get_contract_address(),
+                    self.get_underlying_contract_address()
+                )
+            );
+            self.mutex_account_unlock();
+            return PromiseOrValue::Value(token_amount);
+        }
+
+        mtrading::increase_user_deposit(
+            env::current_account_id(),
+            env::signer_account_id(),
+            token_amount,
+            AccountId::try_from(MARGIN_TRADING_CONTRACT.to_string()).unwrap(),
+            NO_DEPOSIT,
+            self.terra_gas(5),
+        )
+        .then(ext_self::mtrading_increase_user_deposit_callback(
+            env::current_account_id(),
+            env::signer_account_id(),
+            token_amount,
+            self.get_contract_address(),
+            NO_DEPOSIT,
+            self.terra_gas(10),
+        ))
+        .into()
+    }
+
+    #[private]
+    pub fn mtrading_increase_user_deposit_callback(
+        &mut self,
+        amount: WBalance,
+    ) -> PromiseOrValue<U128> {
+        if !is_promise_success() {
+            log!(
+                "{}",
+                Events::MarginTradingFailedToIncreaseUserDeposit(
+                    env::signer_account_id(),
+                    Balance::from(amount)
+                )
+            );
+            self.mutex_account_unlock();
+            return PromiseOrValue::Value(amount);
+        }
+
+        underlying_token::ft_transfer(
```

Comment:
@dmytro-shash, we don't need to transfer tokens to margin trading; it makes no business value in first transferring tokens to the market and then transferring those tokens to another account.

You need to accept the deposit, store the balance somewhere, and provide method to view user balance deposited to the margin trading contract